### DISCUSSION TITLE
File content not updated right away after paste

### DISCUSSION THREAD

**Discussion by justanotheranonymoususer** [microsoft/vscode-discussions#879] [Extension Development QnA]
Hello, I created an issue but it was rejected because the team believed it's a bug in my code, please help me figure out what's the bug if there's any.

My problem is that pasting text, calling `saveAll`, and reading the file quickly sometimes still returns the old content before the save. I saw this especially in a slow VM.

This causes problems for me. I have a setup which does operations on a file with Ctrl+Q. If I press Ctr+V and Ctrl+Q quickly, my workflow reads the old contents, even though I run saveAll!

Below is a deterministic way to reproduce.

Steps to Reproduce:

1. Get [document-editing-sample](https://github.com/Microsoft/vscode-extension-samples/tree/main/document-editing-sample)
2. Replace extension.ts file with this content: https://gist.github.com/justanotheranonymoususer/016f6ada29d6c43487e68fbd066442fe, run `npm i clipboardy@2`, then open the extension folder in VSCode and run it with F5 (clipboardy is used for a simple way to set the clipboard since node.js doesn't have a way to do this natively)
4. Open a new document, type "Hello", and save it as "hello.txt" on the desktop
5. Press Ctrl+Shift+P and run "Reverse Word"
6. Observe an error similar to: `Mismatch, pasted: 6, actual: 5`, meaning that after pasting and saving, the file still contained the text before the paste

Please help me figure out how to overcome this. It's a tricky bug which can cause some sync errors and weird behavior. Thanks.

**Comment by isidorn** [comment]
Can you please link to that bug?

**Comment by justanotheranonymoususer** [comment]
https://github.com/microsoft/vscode/issues/196161

**Comment by tamuratak** [comment]
The following line of your sample just assures that the `paste` event has been triggered. It doesn't assure that the text of the editor has been modified with the content of the clipboard. See the [code](https://github.com/microsoft/vscode/blob/9e597d32277954c4b3259c97fc30c5dc0fb58e13/src/vs/workbench/browser/actions/textInputActions.ts#L51-L53).

```ts
await vscode.commands.executeCommand('editor.action.clipboardPasteAction')
```

After the `paste` event triggered, the listener of the event will eventually modify the text of the editor.

You have to use [onDidChangeTextDocument](https://code.visualstudio.com/api/references/vscode-api#workspace.onDidChangeTextDocument) if you want to do something after the text of the editor modified.

**Reply by justanotheranonymoususer** [reply]
But the real problem is not that I call `editor.action.clipboardPasteAction`, in which case I know paste is pending.

The problem is that the user pressed Ctrl+V and then quickly triggered my action. In this case, I don't know that a paste is pending.

As a user, I expect operations to occur serially, so if I have Ctrl+Q to upload the file to cloud or whatever, that pressing Ctrl+V and then Ctrl+Q will always operate on the pasted content.

**Reply by tamuratak** [reply]
All we can do would be to insert a delay at the beginning of the upload command to wait for the paste to be completed, using `setTimeout`.

**Reply by justanotheranonymoususer** [reply]
That's a horrible solution, it doesn't ensure that the paste will be completed, and it will intorudce a delay in all cases, even if no text was pasted (how can I know?)

That's why I opened a VSCode issue, I think it has to be resolved there, but sadly it was rejected.

**Comment by r3m0t** [comment]
If you use `registerTextEditorCommand`, it will run in sequence with other edits the user makes in the same window.

If you want the VS Code view of the contents of an open file, you should use `vscode.workspace.textDocuments` or `activeTextEditor.document`, that way, you will see the contents in the editor(even if not saved)

If you want the VS Code view of the saved contents, use `vscode.workspace.fs.readFile`, that way, your read will hopefully wait until the save operation properly completed.

Basically, prefer the VS Code API to find information instead of using node functions, and your extension will behave more consistent to VS Code.

**Reply by r3m0t** [reply]
As well, I noticed you logged a difference in length of 1. Make sure this isn't just due to EndOfLine. Some code switches between CRLF and LF.

**Reply by justanotheranonymoususer** [reply]
> If you use `registerTextEditorCommand`, it will run in sequence with other edits the user makes in the same window.

I tried it, but I still can reproduce the issue even if I add a command registered that way and call it after pasting and before saving.

> If you want the VS Code view of the contents of an open file

That's not what I want as my command operates on the file system. Think, executing a third party command line tool.
I can query both the editor content and the file content and compare them, looping while there's a difference, but that's inefficient and not a clean solution.

> If you want the VS Code view of the saved contents

Same as above, I need the file to hold the uptodate content.
Also, just for the test, I replaced `fs.readFileSync` with `vscode.workspace.fs.readFile` and still could reproduce the problem.

> As well, I noticed you logged a difference in length of 1. Make sure this isn't just due to EndOfLine. Some code switches between CRLF and LF.

I'm not sure I understand. There's no newlines in the test.

**Reply by r3m0t** [reply]
Ah you're right. I was able to get your extension to "work" by waiting before calling `saveAll`. I added this to wait until the `TextDocument` was accurate to what you wanted to save-

```typescript

				await new Promise<void>((resolve) => {
					let d = vscode.workspace.onDidChangeTextDocument((e) => {
						if (e.document.uri === editor.document.uri && e.document.getText() === str) {
						 resolve();
						 d.dispose();
						}
					});
					vscode.commands.executeCommand('editor.action.clipboardPasteAction');
				});


				// Save all files
				const saved = await vscode.workspace.saveAll(false);
				if (!saved) {
					vscode.window.showErrorMessage('Could not save files!');
					break;
				}

```

I don't know if that can help your original issue.

**Reply by justanotheranonymoususer** [reply]
Thanks for trying, but as I said above:

> But the real problem is not that I call `editor.action.clipboardPasteAction`, in which case I know paste is pending.
> 
> The problem is that the user pressed Ctrl+V and then quickly triggered my action. In this case, I don't know that a paste is pending.
> 
> As a user, I expect operations to occur serially, so if I have Ctrl+Q to upload the file to cloud or whatever, that pressing Ctrl+V and then Ctrl+Q will always operate on the pasted content.

Also, since you're comparing the whole file content, this comment applies:

> I can query both the editor content and the file content and compare them, looping while there's a difference, but that's inefficient and not a clean solution.

**Comment by theReynald** [comment]
Closing due to inactivity but pls reopen is still in need if help.

**Reply by r3m0t** [reply]
I think this can be closed and kept as an issue and open PR- but why was the issue locked? 
https://github.com/microsoft/vscode/issues/196161
